File: modules/vector-database-adapter.js
/**
 * Vector Database Adapter
 * Provides storage and retrieval for vector embeddings using various backends
 */
const fs = require('fs-extra');
const path = require('path');
const { QdrantClient } = require('@qdrant/js-client-rest');

class VectorDatabaseAdapter {
  /**
   * Create a new vector database adapter
   * @param {Object} options - Configuration options
   * @param {string} options.dbType - Database type ('memory', 'qdrant')
   * @param {number} options.dimensions - Vector dimensions
   * @param {string} options.collectionName - Collection name for storing vectors
   * @param {string} options.qdrantUrl - Qdrant server URL (if using Qdrant)
   * @param {string} options.qdrantApiKey - Qdrant API key (if using cloud Qdrant)
   */
  constructor(options = {}) {
    this.options = Object.assign({
      dbType: 'memory',           // memory, qdrant
      dimensions: 512,            // embedding dimensions
      collectionName: 'canvas_discourse_integration',
      qdrantUrl: 'http://localhost:6333',
      qdrantApiKey: null,
    }, options);
    
    this.client = null;
    this.memoryIndex = {};        // For in-memory storage
    this.initialized = false;
    
    console.log(`Initializing vector database adapter with type: ${this.options.dbType}`);
  }
  
  /**
   * Initialize the database connection
   */
  async initialize() {
    if (this.initialized) return;
    
    try {
      switch(this.options.dbType) {
        case 'qdrant':
          await this.initializeQdrant();
          break;
        case 'memory':
        default:
          this.initializeMemoryIndex();
      }
      
      this.initialized = true;
      console.log(`Vector database initialized with ${this.options.dbType} backend`);
    } catch (error) {
      console.error(`Failed to initialize vector database: ${error.message}`);
      throw error;
    }
  }
  
  /**
   * Initialize in-memory vector storage
   */
  initializeMemoryIndex() {
    this.memoryIndex = {};
    console.log('In-memory vector index initialized');
  }
  
  /**
   * Initialize Qdrant vector database
   */
  async initializeQdrant() {
    try {
      // Create Qdrant client
      const clientOptions = {
        url: this.options.qdrantUrl
      };
      
      if (this.options.qdrantApiKey) {
        clientOptions.apiKey = this.options.qdrantApiKey;
      }
      
      this.client = new QdrantClient(clientOptions);
      
      // Check if collection exists, create if not
      try {
        const collection = await this.client.getCollection(this.options.collectionName);
        console.log(`Using existing Qdrant collection: ${this.options.collectionName}`);
      } catch (error) {
        console.log(`Creating new Qdrant collection: ${this.options.collectionName}`);
        
        await this.client.createCollection(this.options.collectionName, {
          vectors: {
            size: this.options.dimensions,
            distance: 'Cosine'
          },
          optimizers_config: {
            default_segment_number: 2
          },
          on_disk_payload: true
        });
        
        // Create recommended indexes for faster filtering
        await this.client.createPayloadIndex(this.options.collectionName, {
          field_name: 'metadata.system',
          field_schema: 'keyword'
        });
        
        await this.client.createPayloadIndex(this.options.collectionName, {
          field_name: 'metadata.category',
          field_schema: 'keyword'
        });
      }
      
      console.log('Qdrant connection established');
    } catch (error) {
      console.error(`Failed to initialize Qdrant: ${error.message}`);
      throw error;
    }
  }
  
  /**
   * Store a single embedding vector
   * @param {string} id - Document ID
   * @param {Array<number>} vector - Embedding vector
   * @param {Object} metadata - Document metadata
   */
  async storeEmbedding(id, vector, metadata = {}) {
    await this.initialize();
    
    try {
      switch(this.options.dbType) {
        case 'qdrant':
          await this.client.upsert(this.options.collectionName, {
            wait: true,
            points: [{
              id,
              vector,
              payload: { metadata }
            }]
          });
          break;
        case 'memory':
        default:
          this.memoryIndex[id] = { vector, metadata };
      }
      
      console.log(`Stored embedding for document: ${id}`);
      return true;
    } catch (error) {
      console.error(`Failed to store embedding: ${error.message}`);
      return false;
    }
  }
  
  /**
   * Store multiple embedding vectors in batch
   * @param {Array<Object>} embeddings - Array of embeddings to store
   * @param {string} embeddings[].id - Document ID
   * @param {Array<number>} embeddings[].vector - Embedding vector
   * @param {Object} embeddings[].metadata - Document metadata
   */
  async bulkStoreEmbeddings(embeddings) {
    await this.initialize();
    
    if (!embeddings || embeddings.length === 0) {
      console.warn('No embeddings provided for bulk storage');
      return false;
    }
    
    try {
      const batchSize = 100; // Process in batches to avoid overloading
      const batches = [];
      
      // Split into batches
      for (let i = 0; i < embeddings.length; i += batchSize) {
        batches.push(embeddings.slice(i, i + batchSize));
      }
      
      // Process each batch
      for (const batch of batches) {
        switch(this.options.dbType) {
          case 'qdrant':
            await this.client.upsert(this.options.collectionName, {
              wait: true,
              points: batch.map(item => ({
                id: typeof item.id === 'string' ? item.id : String(item.id),
                vector: item.vector,
                payload: { metadata: item.metadata }
              }))
            });
            break;
          case 'memory':
          default:
            batch.forEach(item => {
              this.memoryIndex[item.id] = { 
                vector: item.vector, 
                metadata: item.metadata 
              };
            });
        }
        
        console.log(`Stored batch of ${batch.length} embeddings`);
      }
      
      return true;
    } catch (error) {
      console.error(`Failed to bulk store embeddings: ${error.message}`);
      return false;
    }
  }
  
  /**
   * Import embeddings from a JSON file
   * @param {string} filePath - Path to JSON file with embeddings
   */
  async importEmbeddingsFromFile(filePath) {
    try {
      const fileContent = await fs.readFile(filePath, 'utf8');
      const embeddingsData = JSON.parse(fileContent);
      
      // Convert object format to array format with IDs
      const embeddingsArray = Object.entries(embeddingsData).map(([id, vector]) => ({
        id,
        vector,
        metadata: {
          source: id,
          importedFrom: path.basename(filePath)
        }
      }));
      
      console.log(`Importing ${embeddingsArray.length} embeddings from ${filePath}`);
      
      return await this.bulkStoreEmbeddings(embeddingsArray);
    } catch (error) {
      console.error(`Failed to import embeddings from file: ${error.message}`);
      return false;
    }
  }
  
  /**
   * Find similar documents by vector similarity
   * @param {Array<number>} queryVector - Query embedding vector
   * @param {Object} options - Search options
   * @param {number} options.topK - Number of results to return
   * @param {Object} options.filter - Metadata filters
   * @returns {Array<Object>} Array of similar documents with scores
   */
  async findSimilar(queryVector, options = {}) {
    await this.initialize();
    
    const searchOptions = Object.assign({
      topK: 5,
      filter: {}
    }, options);
    
    try {
      switch(this.options.dbType) {
        case 'qdrant':
          return await this.findSimilarQdrant(queryVector, searchOptions);
        case 'memory':
        default:
          return this.findSimilarMemory(queryVector, searchOptions);
      }
    } catch (error) {
      console.error(`Error during similarity search: ${error.message}`);
      return [];
    }
  }
  
  /**
   * Find similar documents in Qdrant
   */
  async findSimilarQdrant(queryVector, options) {
    // Build filter if needed
    let filter = null;
    
    if (Object.keys(options.filter).length > 0) {
      // Convert simple filters to Qdrant filter format
      filter = { must: [] };
      
      Object.entries(options.filter).forEach(([key, value]) => {
        if (Array.isArray(value)) {
          // Handle array values (OR condition)
          filter.must.push({
            should: value.map(val => ({
              key: `metadata.${key}`,
              match: { value: val }
            }))
          });
        } else {
          // Handle single value
          filter.must.push({
            key: `metadata.${key}`,
            match: { value }
          });
        }
      });
    }
    
    // Query Qdrant
    const result = await this.client.search(this.options.collectionName, {
      vector: queryVector,
      limit: options.topK,
      filter,
      with_payload: true
    });
    
    // Format results to match our standard interface
    return result.map(match => ({
      id: match.id,
      score: match.score,
      metadata: match.payload.metadata
    }));
  }
  
  /**
   * Find similar documents in memory
   */
  findSimilarMemory(queryVector, options) {
    // Calculate cosine similarity for all vectors
    const results = Object.entries(this.memoryIndex).map(([id, entry]) => {
      const similarity = this.cosineSimilarity(queryVector, entry.vector);
      return {
        id,
        score: similarity,
        metadata: entry.metadata
      };
    });
    
    // Apply filters if needed
    let filtered = results;
    if (Object.keys(options.filter).length > 0) {
      filtered = results.filter(item => {
        return Object.entries(options.filter).every(([key, value]) => {
          // Handle array values (OR condition)
          if (Array.isArray(value)) {
            return value.includes(item.metadata[key]);
          }
          // Handle single value
          return item.metadata[key] === value;
        });
      });
    }
    
    // Sort by similarity score (descending) and take top K
    return filtered
      .sort((a, b) => b.score - a.score)
      .slice(0, options.topK);
  }
  
  /**
   * Calculate cosine similarity between two vectors
   */
  cosineSimilarity(vecA, vecB) {
    if (!vecA || !vecB || vecA.length !== vecB.length) return 0;
    
    let dotProduct = 0;
    let normA = 0;
    let normB = 0;
    
    for (let i = 0; i < vecA.length; i++) {
      dotProduct += vecA[i] * vecB[i];
      normA += vecA[i] * vecA[i];
      normB += vecB[i] * vecB[i];
    }
    
    // Handle division by zero
    if (normA === 0 || normB === 0) return 0;
    
    return dotProduct / (Math.sqrt(normA) * Math.sqrt(normB));
  }
  
  /**
   * Delete embeddings from the database
   * @param {string|Array<string>} ids - ID or array of IDs to delete
   */
  async deleteEmbeddings(ids) {
    await this.initialize();
    
    const idArray = Array.isArray(ids) ? ids : [ids];
    
    try {
      switch(this.options.dbType) {
        case 'qdrant':
          await this.client.delete(this.options.collectionName, {
            wait: true,
            points: idArray.map(id => typeof id === 'string' ? id : String(id))
          });
          break;
        case 'memory':
        default:
          idArray.forEach(id => {
            delete this.memoryIndex[id];
          });
      }
      
      console.log(`Deleted ${idArray.length} embeddings`);
      return true;
    } catch (error) {
      console.error(`Failed to delete embeddings: ${error.message}`);
      return false;
    }
  }
  
  /**
   * Clear all embeddings in the database
   */
  async clearAllEmbeddings() {
    await this.initialize();
    
    try {
      switch(this.options.dbType) {
        case 'qdrant':
          // Recreate the collection is the fastest way to clear it
          await this.client.deleteCollection(this.options.collectionName);
          this.initialized = false;
          await this.initialize(); // Reinitialize
          break;
        case 'memory':
        default:
          this.memoryIndex = {};
      }
      
      console.log('Cleared all embeddings');
      return true;
    } catch (error) {
      console.error(`Failed to clear embeddings: ${error.message}`);
      return false;
    }
  }
  
  /**
   * Get database stats
   */
  async getStats() {
    await this.initialize();
    
    try {
      switch(this.options.dbType) {
        case 'qdrant':
          const collInfo = await this.client.getCollection(this.options.collectionName);
          return {
            vectorCount: collInfo.vectors_count,
            status: collInfo.status,
            dimensions: this.options.dimensions,
            dbType: 'qdrant',
            memoryUsage: collInfo.disk_data_size || 0
          };
        case 'memory':
        default:
          return {
            vectorCount: Object.keys(this.memoryIndex).length,
            status: 'active',
            dimensions: this.options.dimensions,
            dbType: 'memory',
            memoryUsage: JSON.stringify(this.memoryIndex).length
          };
      }
    } catch (error) {
      console.error(`Failed to get database stats: ${error.message}`);
      return {
        status: 'error',
        error: error.message
      };
    }
  }
}

module.exports = VectorDatabaseAdapter;

File: src/modules/vector_database_adapter.rs
use std::collections::HashMap;
use std::sync::Arc;
use anyhow::{Result, Context, anyhow};
use async_trait::async_trait;
use log::{info, error};
use serde::{Serialize, Deserialize};
use reqwest::Client;

/// Vector database options
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VectorDatabaseOptions {
    pub db_type: DatabaseType,
    pub dimensions: usize,
    pub collection_name: String,
    pub qdrant_url: String,
    pub qdrant_api_key: Option<String>,
}

impl Default for VectorDatabaseOptions {
    fn default() -> Self {
        Self {
            db_type: DatabaseType::Memory,
            dimensions: 512,
            collection_name: "canvas_discourse_integration".to_string(),
            qdrant_url: "http://localhost:6333".to_string(),
            qdrant_api_key: None,
        }
    }
}

/// Supported database types
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum DatabaseType {
    #[serde(rename = "memory")]
    Memory,
    #[serde(rename = "qdrant")]
    Qdrant,
}

/// Document metadata
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DocumentMetadata {
    #[serde(flatten)]
    pub fields: HashMap<String, String>,
}

/// Document with vector embedding
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Document {
    pub id: String,
    pub vector: Vec<f32>,
    pub metadata: DocumentMetadata,
}

/// Search query result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchResult {
    pub document_id: String,
    pub score: f32,
    pub metadata: DocumentMetadata,
}

/// Search filter
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchFilter {
    pub field: String,
    pub value: String,
}

/// Qdrant point payload
#[derive(Debug, Clone, Serialize, Deserialize)]
struct QdrantPayload {
    metadata: DocumentMetadata,
}

/// Qdrant point
#[derive(Debug, Clone, Serialize, Deserialize)]
struct QdrantPoint {
    id: String,
    vector: Vec<f32>,
    payload: QdrantPayload,
}

/// Qdrant search request
#[derive(Debug, Clone, Serialize, Deserialize)]
struct QdrantSearchRequest {
    vector: Vec<f32>,
    limit: usize,
    filter: Option<QdrantFilter>,
    with_payload: bool,
}

/// Qdrant filter
#[derive(Debug, Clone, Serialize, Deserialize)]
struct QdrantFilter {
    should: Vec<QdrantCondition>,
}

/// Qdrant filter condition
#[derive(Debug, Clone, Serialize, Deserialize)]
struct QdrantCondition {
    key: String,
    match_value: String,
}

/// Qdrant collection creation request
#[derive(Debug, Clone, Serialize, Deserialize)]
struct QdrantCreateCollectionRequest {
    vectors: QdrantVectorConfig,
    optimizers_config: QdrantOptimizerConfig,
    on_disk_payload: bool,
}

/// Qdrant vector configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
struct QdrantVectorConfig {
    size: usize,
    distance: String,
}

/// Qdrant optimizer configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
struct QdrantOptimizerConfig {
    default_segment_number: usize,
}

/// Qdrant index creation request
#[derive(Debug, Clone, Serialize, Deserialize)]
struct QdrantCreateIndexRequest {
    field_name: String,
    field_schema: String,
}

/// Qdrant upsert request
#[derive(Debug, Clone, Serialize, Deserialize)]
struct QdrantUpsertRequest {
    wait: bool,
    points: Vec<QdrantPoint>,
}

/// Qdrant search response
#[derive(Debug, Clone, Serialize, Deserialize)]
struct QdrantSearchResponse {
    result: Vec<QdrantSearchResult>,
}

/// Qdrant search result
#[derive(Debug, Clone, Serialize, Deserialize)]
struct QdrantSearchResult {
    id: String,
    score: f32,
    payload: QdrantPayload,
}

/// Vector database adapter trait
#[async_trait]
pub trait VectorDatabase: Send + Sync {
    /// Initialize the database
    async fn initialize(&mut self) -> Result<()>;
    
    /// Store an embedding vector
    async fn store_embedding(
        &self,
        id: &str,
        vector: Vec<f32>,
        metadata: DocumentMetadata,
    ) -> Result<bool>;
    
    /// Store multiple embedding vectors
    async fn store_embeddings(
        &self,
        documents: Vec<Document>,
    ) -> Result<bool>;
    
    /// Find similar vectors
    async fn find_similar(
        &self,
        vector: &[f32],
        limit: usize,
        filter: Option<SearchFilter>,
    ) -> Result<Vec<SearchResult>>;
    
    /// Delete a document
    async fn delete_document(&self, id: &str) -> Result<bool>;
    
    /// Get document by ID
    async fn get_document(&self, id: &str) -> Result<Option<Document>>;
}

/// Memory-based vector database
pub struct MemoryVectorDatabase {
    memory_index: HashMap<String, Document>,
    initialized: bool,
}

impl MemoryVectorDatabase {
    /// Create a new memory-based vector database
    pub fn new() -> Self {
        Self {
            memory_index: HashMap::new(),
            initialized: false,
        }
    }
}

#[async_trait]
impl VectorDatabase for MemoryVectorDatabase {
    async fn initialize(&mut self) -> Result<()> {
        if self.initialized {
            return Ok(());
        }
        
        self.memory_index = HashMap::new();
        info!("In-memory vector index initialized");
        
        self.initialized = true;
        Ok(())
    }
    
    async fn store_embedding(
        &self,
        id: &str,
        vector: Vec<f32>,
        metadata: DocumentMetadata,
    ) -> Result<bool> {
        let mut memory_index = self.memory_index.clone();
        
        memory_index.insert(
            id.to_string(),
            Document {
                id: id.to_string(),
                vector,
                metadata,
            },
        );
        
        info!("Stored embedding for document: {}", id);
        Ok(true)
    }
    
    async fn store_embeddings(
        &self,
        documents: Vec<Document>,
    ) -> Result<bool> {
        let mut memory_index = self.memory_index.clone();
        
        for doc in documents {
            memory_index.insert(doc.id.clone(), doc);
        }
        
        info!("Stored {} embeddings in memory", documents.len());
        Ok(true)
    }
    
    async fn find_similar(
        &self,
        vector: &[f32],
        limit: usize,
        filter: Option<SearchFilter>,
    ) -> Result<Vec<SearchResult>> {
        let mut results: Vec<(String, f32, DocumentMetadata)> = Vec::new();
        
        // Calculate cosine similarity for each vector in memory
        for (id, doc) in &self.memory_index {
            // Apply filter if specified
            if let Some(filter_condition) = &filter {
                if let Some(field_value) = doc.metadata.fields.get(&filter_condition.field) {
                    if field_value != &filter_condition.value {
                        continue;
                    }
                } else {
                    continue;
                }
            }
            
            let similarity = cosine_similarity(vector, &doc.vector);
            results.push((id.clone(), similarity, doc.metadata.clone()));
        }
        
        // Sort by similarity (descending)
        results.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));
        
        // Take top 'limit' results
        let search_results: Vec<SearchResult> = results
            .iter()
            .take(limit)
            .map(|(id, score, metadata)| SearchResult {
                document_id: id.clone(),
                score: *score,
                metadata: metadata.clone(),
            })
            .collect();
        
        Ok(search_results)
    }
    
    async fn delete_document(&self, id: &str) -> Result<bool> {
        let mut memory_index = self.memory_index.clone();
        
        memory_index.remove(id);
        
        info!("Deleted document: {}", id);
        Ok(true)
    }
    
    async fn get_document(&self, id: &str) -> Result<Option<Document>> {
        Ok(self.memory_index.get(id).cloned())
    }
}

/// Qdrant vector database
pub struct QdrantVectorDatabase {
    client: Client,
    options: VectorDatabaseOptions,
    initialized: bool,
}

impl QdrantVectorDatabase {
    /// Create a new Qdrant vector database
    pub fn new(options: VectorDatabaseOptions) -> Self {
        Self {
            client: Client::new(),
            options,
            initialized: false,
        }
    }
    
    /// Get the base URL for Qdrant API
    fn base_url(&self) -> String {
        format!("{}/collections/{}", self.options.qdrant_url, self.options.collection_name)
    }
    
    /// Build a request with authentication if needed
    fn build_request(&self, url: &str) -> reqwest::RequestBuilder {
        let mut builder = self.client.get(url);
        
        if let Some(api_key) = &self.options.qdrant_api_key {
            builder = builder.header("api-key", api_key);
        }
        
        builder
    }
}

#[async_trait]
impl VectorDatabase for QdrantVectorDatabase {
    async fn initialize(&mut self) -> Result<()> {
        if self.initialized {
            return Ok(());
        }
        
        // Check if collection exists
        let collection_url = format!("{}/collections/{}", 
            self.options.qdrant_url, 
            self.options.collection_name
        );
        
        let mut builder = self.client.get(&collection_url);
        
        if let Some(api_key) = &self.options.qdrant_api_key {
            builder = builder.header("api-key", api_key);
        }
        
        let response = builder.send().await;
        
        match response {
            Ok(resp) if resp.status().is_success() => {
                info!("Using existing Qdrant collection: {}", self.options.collection_name);
            }
            _ => {
                // Create new collection
                info!("Creating new Qdrant collection: {}", self.options.collection_name);
                
                let create_request = QdrantCreateCollectionRequest {
                    vectors: QdrantVectorConfig {
                        size: self.options.dimensions,
                        distance: "Cosine".to_string(),
                    },
                    optimizers_config: QdrantOptimizerConfig {
                        default_segment_number: 2,
                    },
                    on_disk_payload: true,
                };
                
                let mut builder = self.client.put(&collection_url);
                
                if let Some(api_key) = &self.options.qdrant_api_key {
                    builder = builder.header("api-key", api_key);
                }
                
                let response = builder
                    .json(&create_request)
                    .send()
                    .await
                    .context("Failed to create Qdrant collection")?;
                    
                if !response.status().is_success() {
                    let error_text = response.text().await.unwrap_or_default();
                    return Err(anyhow!("Failed to create Qdrant collection: {}", error_text));
                }
                
                // Create recommended indexes
                let index_url = format!("{}/index", &collection_url);
                
                for field in &["metadata.system", "metadata.category"] {
                    let index_request = QdrantCreateIndexRequest {
                        field_name: field.to_string(),
                        field_schema: "keyword".to_string(),
                    };
                    
                    let mut builder = self.client.put(&index_url);
                    
                    if let Some(api_key) = &self.options.qdrant_api_key {
                        builder = builder.header("api-key", api_key);
                    }
                    
                    let response = builder
                        .json(&index_request)
                        .send()
                        .await
                        .context(format!("Failed to create index for field: {}", field))?;
                        
                    if !response.status().is_success() {
                        let error_text = response.text().await.unwrap_or_default();
                        return Err(anyhow!("Failed to create index: {}", error_text));
                    }
                }
            }
        }
        
        info!("Qdrant connection established");
        self.initialized = true;
        Ok(())
    }
    
    async fn store_embedding(
        &self,
        id: &str,
        vector: Vec<f32>,
        metadata: DocumentMetadata,
    ) -> Result<bool> {
        let points_url = format!("{}/points", self.base_url());
        
        let upsert_request = QdrantUpsertRequest {
            wait: true,
            points: vec![
                QdrantPoint {
                    id: id.to_string(),
                    vector,
                    payload: QdrantPayload { metadata },
                }
            ],
        };
        
        let mut builder = self.client.put(&points_url);
        
        if let Some(api_key) = &self.options.qdrant_api_key {
            builder = builder.header("api-key", api_key);
        }
        
        let response = builder
            .json(&upsert_request)
            .send()
            .await
            .context("Failed to store embedding in Qdrant")?;
            
        if !response.status().is_success() {
            let error_text = response.text().await.unwrap_or_default();
            return Err(anyhow!("Failed to store embedding: {}", error_text));
        }
        
        info!("Stored embedding for document: {}", id);
        Ok(true)
    }
    
    async fn store_embeddings(
        &self,
        documents: Vec<Document>,
    ) -> Result<bool> {
        let points_url = format!("{}/points", self.base_url());
        
        let qdrant_points: Vec<QdrantPoint> = documents
            .into_iter()
            .map(|doc| QdrantPoint {
                id: doc.id,
                vector: doc.vector,
                payload: QdrantPayload { metadata: doc.metadata },
            })
            .collect();
        
        let upsert_request = QdrantUpsertRequest {
            wait: true,
            points: qdrant_points,
        };
        
        let mut builder = self.client.put(&points_url);
        
        if let Some(api_key) = &self.options.qdrant_api_key {
            builder = builder.header("api-key", api_key);
        }
        
        let response = builder
            .json(&upsert_request)
            .send()
            .await
            .context("Failed to store embeddings in Qdrant")?;
            
        if !response.status().is_success() {
            let error_text = response.text().await.unwrap_or_default();
            return Err(anyhow!("Failed to store embeddings: {}", error_text));
        }
        
        info!("Stored {} embeddings in Qdrant", upsert_request.points.len());
        Ok(true)
    }
    
    async fn find_similar(
        &self,
        vector: &[f32],
        limit: usize,
        filter: Option<SearchFilter>,
    ) -> Result<Vec<SearchResult>> {
        let search_url = format!("{}/points/search", self.base_url());
        
        let qdrant_filter = filter.map(|f| QdrantFilter {
            should: vec![
                QdrantCondition {
                    key: f.field,
                    match_value: f.value,
                }
            ],
        });
        
        let search_request = QdrantSearchRequest {
            vector: vector.to_vec(),
            limit,
            filter: qdrant_filter,
            with_payload: true,
        };
        
        let mut builder = self.client.post(&search_url);
        
        if let Some(api_key) = &self.options.qdrant_api_key {
            builder = builder.header("api-key", api_key);
        }
        
        let response = builder
            .json(&search_request)
            .send()
            .await
            .context("Failed to search in Qdrant")?;
            
        if !response.status().is_success() {
            let error_text = response.text().await.unwrap_or_default();
            return Err(anyhow!("Failed to search: {}", error_text));
        }
        
        let search_response: QdrantSearchResponse = response
            .json()
            .await
            .context("Failed to parse Qdrant search response")?;
            
        let results = search_response.result
            .into_iter()
            .map(|r| SearchResult {
                document_id: r.id,
                score: r.score,
                metadata: r.payload.metadata,
            })
            .collect();
            
        Ok(results)
    }
    
    async fn delete_document(&self, id: &str) -> Result<bool> {
        let delete_url = format!("{}/points/delete", self.base_url());
        
        let mut builder = self.client.post(&delete_url);
        
        if let Some(api_key) = &self.options.qdrant_api_key {
            builder = builder.header("api-key", api_key);
        }
        
        let response = builder
            .json(&serde_json::json!({
                "points": [id],
                "wait": true
            }))
            .send()
            .await
            .context("Failed to delete document from Qdrant")?;
            
        if !response.status().is_success() {
            let error_text = response.text().await.unwrap_or_default();
            return Err(anyhow!("Failed to delete document: {}", error_text));
        }
        
        info!("Deleted document: {}", id);
        Ok(true)
    }
    
    async fn get_document(&self, id: &str) -> Result<Option<Document>> {
        let get_url = format!("{}/points/{}", self.base_url(), id);
        
        let mut builder = self.client.get(&get_url);
        
        if let Some(api_key) = &self.options.qdrant_api_key {
            builder = builder.header("api-key", api_key);
        }
        
        let response = builder
            .send()
            .await
            .context("Failed to get document from Qdrant")?;
            
        if response.status().is_client_error() {
            return Ok(None);
        }
        
        if !response.status().is_success() {
            let error_text = response.text().await.unwrap_or_default();
            return Err(anyhow!("Failed to get document: {}", error_text));
        }
        
        #[derive(Deserialize)]
        struct QdrantGetResponse {
            result: QdrantPoint,
        }
        
        let get_response: QdrantGetResponse = response
            .json()
            .await
            .context("Failed to parse Qdrant get response")?;
            
        let point = get_response.result;
        
        Ok(Some(Document {
            id: point.id,
            vector: point.vector,
            metadata: point.payload.metadata,
        }))
    }
}

/// Vector Database Adapter
/// Provides storage and retrieval for vector embeddings using various backends
pub struct VectorDatabaseAdapter {
    options: VectorDatabaseOptions,
    db: Option<Arc<dyn VectorDatabase>>,
    initialized: bool,
}

impl VectorDatabaseAdapter {
    /// Create a new vector database adapter
    pub fn new(options: Option<VectorDatabaseOptions>) -> Self {
        let options = options.unwrap_or_default();
        
        info!("Initializing vector database adapter with type: {:?}", options.db_type);
        
        Self {
            options,
            db: None,
            initialized: false,
        }
    }
    
    /// Initialize the database connection
    pub async fn initialize(&mut self) -> Result<()> {
        if self.initialized {
            return Ok(());
        }
        
        match self.options.db_type {
            DatabaseType::Qdrant => {
                let mut db = QdrantVectorDatabase::new(self.options.clone());
                db.initialize().await?;
                self.db = Some(Arc::new(db));
            }
            DatabaseType::Memory => {
                let mut db = MemoryVectorDatabase::new();
                db.initialize().await?;
                self.db = Some(Arc::new(db));
            }
        }
        
        self.initialized = true;
        info!("Vector database initialized with {:?} backend", self.options.db_type);
        
        Ok(())
    }
    
    /// Store a single embedding vector
    pub async fn store_embedding(
        &self,
        id: &str,
        vector: Vec<f32>,
        metadata: HashMap<String, String>,
    ) -> Result<bool> {
        self.ensure_initialized().await?;
        
        let db = self.db.as_ref().unwrap();
        
        db.store_embedding(
            id,
            vector,
            DocumentMetadata { fields: metadata },
        ).await
    }
    
    /// Store multiple embedding vectors
    pub async fn store_embeddings(
        &self,
        documents: Vec<(String, Vec<f32>, HashMap<String, String>)>,
    ) -> Result<bool> {
        self.ensure_initialized().await?;
        
        let db = self.db.as_ref().unwrap();
        
        let docs: Vec<Document> = documents
            .into_iter()
            .map(|(id, vector, metadata)| Document {
                id,
                vector,
                metadata: DocumentMetadata { fields: metadata },
            })
            .collect();
            
        db.store_embeddings(docs).await
    }
    
    /// Find similar vectors
    pub async fn find_similar(
        &self,
        vector: &[f32],
        options: HashMap<String, serde_json::Value>,
    ) -> Result<Vec<SearchResult>> {
        self.ensure_initialized().await?;
        
        let db = self.db.as_ref().unwrap();
        
        // Extract options
        let limit = options
            .get("topK")
            .and_then(|v| v.as_u64())
            .unwrap_or(5) as usize;
            
        let filter = if let Some(filter_obj) = options.get("filter") {
            if let Some(obj) = filter_obj.as_object() {
                // Just use the first filter for simplicity
                if let Some((field, value)) = obj.iter().next() {
                    if let Some(value_str) = value.as_str() {
                        Some(SearchFilter {
                            field: field.clone(),
                            value: value_str.to_string(),
                        })
                    } else {
                        None
                    }
                } else {
                    None
                }
            } else {
                None
            }
        } else {
            None
        };
        
        db.find_similar(vector, limit, filter).await
    }
    
    /// Delete a document
    pub async fn delete_document(&self, id: &str) -> Result<bool> {
        self.ensure_initialized().await?;
        
        let db = self.db.as_ref().unwrap();
        db.delete_document(id).await
    }
    
    /// Get document by ID
    pub async fn get_document(&self, id: &str) -> Result<Option<Document>> {
        self.ensure_initialized().await?;
        
        let db = self.db.as_ref().unwrap();
        db.get_document(id).await
    }
    
    /// Ensure the database is initialized
    async fn ensure_initialized(&self) -> Result<()> {
        if !self.initialized {
            return Err(anyhow!("Vector database not initialized. Call initialize() first."));
        }
        
        Ok(())
    }
}

/// Calculate cosine similarity between two vectors
fn cosine_similarity(a: &[f32], b: &[f32]) -> f32 {
    let mut dot_product = 0.0;
    let mut norm_a = 0.0;
    let mut norm_b = 0.0;
    
    for i in 0..a.len().min(b.len()) {
        dot_product += a[i] * b[i];
        norm_a += a[i] * a[i];
        norm_b += b[i] * b[i];
    }
    
    if norm_a == 0.0 || norm_b == 0.0 {
        return 0.0;
    }
    
    dot_product / (norm_a.sqrt() * norm_b.sqrt())
}
